File: benchmark/python/arenabench.py
"""Arena-stress benchmarks (group `arena`).

Mirrors benchmark/mfb/src/arena.mfb: mixed-size transient churn, long+short-lived
mix, and grow/shrink. These are the mfb runtime's plan-39-A regression gate; the
Python allocator is unaffected, so these mirrors keep the same tiny counts only
so the table lines up. Native Python list/str reproduce the exact arithmetic, so
the checksums match the mfb reference.
"""
import sys

RUN = 1
now_ns = None
record = None


def test_arena_transient():
    base = "abcdefghij"
    times = []
    checksum = 0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0
        for i in range(400):
            size = (i % 16) + 1
            tmp = []
            for k in range(size):
                tmp.append(k)
            acc += sum(tmp)
            slicelen = (i % 7) + 1
            s = base[:slicelen]
            acc += len(s)
        checksum = acc
        times.append(now_ns() - t0)
    print("arena_transient = %d" % checksum, file=sys.stderr)
    record("arena", "transient", times)


def test_arena_mixed():
    long_lived = []
    for i in range(1000):
        long_lived.append(i)
    times = []
    checksum = 0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0
        for i in range(400):
            size = (i % 20) + 1
            tmp = []
            for k in range(size):
                tmp.append(k)
            acc += sum(tmp)
            s = "ab" * ((i % 5) + 1)
            acc += len(s)
        acc += len(long_lived)
        checksum = acc
        times.append(now_ns() - t0)
    print("arena_mixed = %d" % checksum, file=sys.stderr)
    record("arena", "mixed", times)


def test_arena_growshrink():
    grow = 100
    cycles = 200
    times = []
    checksum = 0
    for _ in range(RUN):
        t0 = now_ns()
        acc = 0
        for _cycle in range(cycles):
            xs = []
            for k in range(grow):
                xs.append(k)
            head = xs[:10]
            tail = xs[grow - 10:]
            acc += len(head) + len(tail)
        checksum = acc
        times.append(now_ns() - t0)
    print("arena_growshrink = %d" % checksum, file=sys.stderr)
    record("arena", "growshrink", times)


def run_all(run, now_ns_fn, record_fn):
    global RUN, now_ns, record
    RUN, now_ns, record = run, now_ns_fn, record_fn
    test_arena_transient()
    test_arena_mixed()
    test_arena_growshrink()

File: benchmark/c/arenabench.c
/* GROUP: arena — the C oracle for arena.mfb. C's allocator has no arena free-
 * list degradation, so these just reproduce the mfb loop arithmetic exactly
 * (with real malloc'd temporaries for comparable work). Checksums must match:
 *   transient=18597, mixed=30000, growshrink=4000 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bench.h"
#include "arenabench.h"

static void test_arena_transient(void) {
  const char *base = "abcdefghij"; /* len 10 */
  long long *t = alloc_times();
  long checksum = 0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    long acc = 0;
    for (int i = 0; i < 400; i++) {
      int size = (i % 16) + 1;
      long *tmp = malloc((size_t)size * sizeof(long));
      for (int k = 0; k < size; k++) tmp[k] = k;
      long s = 0;
      for (int k = 0; k < size; k++) s += tmp[k];
      acc += s; /* collections::sum(tmp) */
      free(tmp);
      int slicelen = (i % 7) + 1;
      char *sl = malloc((size_t)slicelen + 1);
      memcpy(sl, base, (size_t)slicelen);
      sl[slicelen] = '\0';
      acc += (long)strlen(sl); /* len(strings::left(base, slicelen)) */
      free(sl);
    }
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "arena_transient = %ld\n", checksum);
  record("arena", "transient", t, RUN);
  free(t);
}

static void test_arena_mixed(void) {
  long *longLived = malloc(1000 * sizeof(long));
  for (int i = 0; i < 1000; i++) longLived[i] = i;
  long long *t = alloc_times();
  long checksum = 0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    long acc = 0;
    for (int i = 0; i < 400; i++) {
      int size = (i % 20) + 1;
      long *tmp = malloc((size_t)size * sizeof(long));
      for (int k = 0; k < size; k++) tmp[k] = k;
      long s = 0;
      for (int k = 0; k < size; k++) s += tmp[k];
      acc += s;
      free(tmp);
      int reps = (i % 5) + 1; /* strings::repeat("ab", reps) */
      char *str = malloc((size_t)reps * 2 + 1);
      for (int k = 0; k < reps; k++) memcpy(str + (size_t)k * 2, "ab", 2);
      str[reps * 2] = '\0';
      acc += (long)strlen(str);
      free(str);
    }
    acc += 1000; /* len(longLived) */
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "arena_mixed = %ld\n", checksum);
  record("arena", "mixed", t, RUN);
  free(t);
  free(longLived);
}

static void test_arena_growshrink(void) {
  const int grow = 100, cycles = 200;
  long long *t = alloc_times();
  long checksum = 0;
  for (int r = 0; r < RUN; r++) {
    long long t0 = now_ns();
    long acc = 0;
    for (int cycle = 0; cycle < cycles; cycle++) {
      long *xs = malloc((size_t)grow * sizeof(long));
      for (int k = 0; k < grow; k++) xs[k] = k;
      int headn = 10 < grow ? 10 : grow;               /* take(xs, 10) */
      int tailn = grow - (grow - 10);                  /* drop(xs, grow-10) */
      acc += headn + tailn;
      free(xs);
    }
    checksum = acc;
    t[r] = now_ns() - t0;
  }
  fprintf(stderr, "arena_growshrink = %ld\n", checksum);
  record("arena", "growshrink", t, RUN);
  free(t);
}

void run_arena_group(void) {
  test_arena_transient();
  test_arena_mixed();
  test_arena_growshrink();
}
